File: apps/backend/src/modules/announcement/entities/announcement.entity.ts
import {
  Entity,
  Column,
  Index,
  ManyToOne,
  OneToMany,
  JoinColumn,
} from 'typeorm';
import { TenantBaseEntity } from '../../../shared/database/tenant-base.entity';
import { AnnouncementCategory } from '../enums/announcement-category.enum';
import { AnnouncementPriority } from '../enums/announcement-priority.enum';
import { AnnouncementTargetAudience } from '../enums/announcement-target-audience.enum';
import { AnnouncementRead } from './announcement-read.entity';
import { User } from '../../iam/entities/user.entity';

@Entity('announcements')
@Index(['companyId', 'isPublished', 'expiresAt'])
@Index(['companyId', 'scheduledAt'])
@Index(['companyId', 'category'])
@Index(['companyId', 'priority'])
export class Announcement extends TenantBaseEntity {
  @Column({ type: 'uuid', name: 'created_by_user_id' })
  createdByUserId!: string;

  @ManyToOne(() => User, { nullable: false })
  @JoinColumn({ name: 'created_by_user_id' })
  createdByUser!: User;

  @Column({ type: 'text' })
  title!: string;

  @Column({ type: 'text' })
  message!: string;

  @Column({
    type: 'enum',
    enum: AnnouncementCategory,
    default: AnnouncementCategory.GENERAL,
  })
  category!: AnnouncementCategory;

  @Column({
    type: 'enum',
    enum: AnnouncementPriority,
    default: AnnouncementPriority.MEDIUM,
  })
  priority!: AnnouncementPriority;

  @Column({
    type: 'enum',
    enum: AnnouncementTargetAudience,
    name: 'target_audience',
  })
  targetAudience!: AnnouncementTargetAudience;

  @Column({ type: 'jsonb', name: 'target_roles', nullable: true })
  targetRoles!: string[] | null;

  @Column({ type: 'timestamptz', name: 'scheduled_at', nullable: true })
  scheduledAt!: Date | null;

  @Column({ type: 'timestamptz', name: 'expires_at', nullable: true })
  expiresAt!: Date | null;

  @Column({ type: 'boolean', name: 'is_published', default: false })
  isPublished!: boolean;

  @Column({ type: 'timestamptz', name: 'published_at', nullable: true })
  publishedAt!: Date | null;

  @Column({ type: 'jsonb', nullable: true })
  metadata!: Record<string, unknown> | null;

  @Column({ type: 'timestamptz', name: 'deleted_at', nullable: true })
  deletedAt!: Date | null;

  @OneToMany(() => AnnouncementRead, (read) => read.announcement, {
    cascade: true,
  })
  reads!: AnnouncementRead[];
}

File: apps/frontend/lib/src/features/announcement/domain/entities/announcement_entity.dart
import 'package:equatable/equatable.dart';

enum AnnouncementCategory {
  maintenance,
  emergency,
  general,
  info;

  static AnnouncementCategory fromString(String value) {
    return AnnouncementCategory.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => AnnouncementCategory.general,
    );
  }

  String get displayName {
    switch (this) {
      case AnnouncementCategory.maintenance:
        return 'Maintenance';
      case AnnouncementCategory.emergency:
        return 'Emergency';
      case AnnouncementCategory.general:
        return 'General';
      case AnnouncementCategory.info:
        return 'Info';
    }
  }
}

enum AnnouncementPriority {
  low,
  medium,
  high,
  urgent;

  static AnnouncementPriority fromString(String value) {
    return AnnouncementPriority.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => AnnouncementPriority.medium,
    );
  }

  String get displayName {
    switch (this) {
      case AnnouncementPriority.low:
        return 'Low';
      case AnnouncementPriority.medium:
        return 'Medium';
      case AnnouncementPriority.high:
        return 'High';
      case AnnouncementPriority.urgent:
        return 'Urgent';
    }
  }
}

enum AnnouncementTargetAudience {
  all,
  roles;

  static AnnouncementTargetAudience fromString(String value) {
    return AnnouncementTargetAudience.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => AnnouncementTargetAudience.all,
    );
  }
}

class AnnouncementEntity extends Equatable {
  const AnnouncementEntity({
    required this.id,
    required this.companyId,
    required this.createdByUserId,
    required this.title,
    required this.message,
    required this.category,
    required this.priority,
    required this.targetAudience,
    this.targetRoles,
    this.scheduledAt,
    this.expiresAt,
    required this.isPublished,
    this.publishedAt,
    this.metadata,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String companyId;
  final String createdByUserId;
  final String title;
  final String message;
  final AnnouncementCategory category;
  final AnnouncementPriority priority;
  final AnnouncementTargetAudience targetAudience;
  final List<String>? targetRoles;
  final DateTime? scheduledAt;
  final DateTime? expiresAt;
  final bool isPublished;
  final DateTime? publishedAt;
  final Map<String, dynamic>? metadata;
  final DateTime createdAt;
  final DateTime updatedAt;

  @override
  List<Object?> get props => [
        id,
        companyId,
        createdByUserId,
        title,
        message,
        category,
        priority,
        targetAudience,
        targetRoles,
        scheduledAt,
        expiresAt,
        isPublished,
        publishedAt,
        metadata,
        createdAt,
        updatedAt,
      ];
}

